zerver/lib/streams.py

```diff
@@ -721,10 +722,24 @@ def can_access_stream_user_ids(stream: Stream) -> set[int]:
         # except unsubscribed guest users
         return public_stream_user_ids(stream)
     else:
-        # for a private stream, it's subscribers plus realm admins.
-        return private_stream_user_ids(stream.id) | {
-            user.id for user in stream.realm.get_admin_users_and_bots()
-        }
+        # for a private stream, it's subscribers plus channel admins
+        # and users belonging to `can_add_subscribers_group`.
+        stream_admin_user_ids = set(
+            get_recursive_group_members(stream.can_administer_channel_group).values_list(
+                "id", flat=True
+            )
+        )
+        stream_add_subscribers_group_user_ids = set(
+            get_recursive_group_members(stream.can_add_subscribers_group).values_list(
+                "id", flat=True
+            )
+        )
```

Comment:
Would it be helpful to have a helper function that, given a list of groups, returns the union of their memberships? I'm kinda surprised this commit didn't add extra queries.